### PR TITLE
New principle: Avoid introducing new parser-blocking features, fixes #329

### LINKED ISSUE #329: New principle: Avoid introducing new parser-blocking features
This was brought up here: https://github.com/whatwg/html/issues/2791#issuecomment-892790934
and I thought it's exactly the kind of thing we should have a principle about.

### PR DESCRIPTION
**Pull request by LeaVerou**
As discussed in #329

Text by @jakearchibald

***
<a href="https://pr-preview.s3.amazonaws.com/w3ctag/design-principles/pull/347.html" title="Last updated on Dec 9, 2021, 8:54 AM UTC (8808b38)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3ctag/design-principles/347/07386d3...8808b38.html" title="Last updated on Dec 9, 2021, 8:54 AM UTC (8808b38)">Diff</a>

***
<a href="https://pr-preview.s3.amazonaws.com/w3ctag/design-principles/pull/347.html" title="Last updated on Jan 24, 2022, 5:57 PM UTC (aa0d5a5)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3ctag/design-principles/347/07386d3...aa0d5a5.html" title="Last updated on Jan 24, 2022, 5:57 PM UTC (aa0d5a5)">Diff</a>

### CODE REVIEW

**Comment by torgo** [comment]
We discussed this morning and feel it should be compressed down to one principle and made a bit clearer.  Sangwhan is going to take a whack at it.

**Comment by jakearchibald** [comment]
I'd advise not conflating parser blocking with render blocking since the latter is ok in very particular situations.

File: index.bs [hunk 1]
```diff
@@ -848,6 +848,36 @@ try to keep its syntax as close as possible to the syntax of the existing attrib
...
 </div>
 
+## Do not pause the HTML parser
+
+Ensure that your design does not require HTML parser to pause to handle external resources.
+
+As a page parses,
+the browser discovers assets that the page needs,
+and figures out a priority in which they should be loaded in parallel.
+This is disrupted by a resource which blocks the discovery of subsequent resources.
```
Comment (atanassov) [inline index.bs:858]:
```suggestion
Such parsing can be disrupted by a resource which blocks the discovery of subsequent resources.
```

File: index.bs [hunk 2]
```diff
@@ -848,6 +848,36 @@ try to keep its syntax as close as possible to the syntax of the existing attrib
 	and the other one a selector, that would be an antipattern.
 </div>
 
+## Do not pause the HTML parser
```
Comment (atanassov) [inline index.bs:851]:
```suggestion
## Avoid pausing of the HTML parser
```

File: index.bs [hunk 3]
```diff
@@ -848,6 +848,36 @@ try to keep its syntax as close as possible to the syntax of the existing attrib
...
+Features that block the parser generally do so
+because they want to feed additional content into the parser
+before subsequent content.
+This is the case of legacy `<script src="…">` elements,
+which can inject into the parser using `document.write(…)`.
+Due to the performance issues above, new features must not do this.
+
+## Be cautious about blocking the renderer
+
+Block rendering only if it improves user experience.
```
Comment (atanassov) [inline index.bs:872]:
```suggestion
## Avoid features that block rendering

Features that block rendering should be considered only if user experience is improved.
```

File: index.bs [hunk 4]
```diff
@@ -848,6 +848,36 @@ try to keep its syntax as close as possible to the syntax of the existing attrib
...
+
+If the browser needs to download an additional resource before rendering the page,
+it extends the time the user is looking at a blank page (or the previous page).
+
+Features should only do this in order to prevent a worse user experience.
+For instance, blocking rendering on a stylesheet download makes sense,
+as the alternative is a flash of unstyled content.
```
Comment (atanassov) [inline index.bs:879]:
```suggestion
Features that require resource loading or other operations before rendering the page,
often result in blank page (or the previous page). The result is a poor user experience.

Consider adding such features only in cases when the overall user experience is improved.
A canonical example of this is blocking rendering in order to download and process stylesheet. 
The alternative user experience is a flash of unstyled content, which is undesirable.
```

File: index.bs (hunk 3 above)
Comment (jakearchibald) [reply index.bs:872]: "Avoid pausing the HTML parser"
"Avoid features that block rendering"

At a glance, the wording here suggests both rules carry equal weight, but pausing the parser is much worse than blocking rendering.

Comment (atanassov) [reply index.bs:872]: That's fair. Render blocking is certainly a subset of parser blocking in terms of badness. Render blocking is also more combinatorial compared to blocking the HTML parser.

I'll revert the HTML parser section heading so it holds stronger.

File: index.bs [hunk 5]
```diff
@@ -848,6 +848,34 @@ try to keep its syntax as close as possible to the syntax of the existing attrib
 	and the other one a selector, that would be an antipattern.
 </div>
 
+## Do not pause the HTML parser
```
Comment (atanassov) [inline index.bs:851]:
```suggestion
<h3 id="avoid-html-parser-blocking"> Do not pause the HTML parser</h3>
```

File: index.bs [hunk 6]
```diff
@@ -848,6 +848,34 @@ try to keep its syntax as close as possible to the syntax of the existing attrib
...
+which may turn out to be incorrect.
+
+Features that block the parser generally do so
+because they want to feed additional content into the parser
+before subsequent content.
+This is the case of legacy `<script src="…">` elements,
+which can inject into the parser using `document.write(…)`.
+Due to the performance issues above, new features must not do this.
+
+## Avoid features that block rendering
```
Comment (atanassov) [inline index.bs:870]:
```suggestion
<h3 id="avoid-reder-blocking"> Avoid features that block rendering</h3>
```

**Review by plinss — APPROVED** [review]

**Pull request merged**